### PR TITLE
[#36] allow changing almost everything in Config using System Properties

### PR DESCRIPTION
**Pull request by nezda**
[#36] allow changing almost everything in Config using System Properties
* note: used underscores instead of periods used in the joni.debug* options
* had to tweak `Search.java` for the change to compile

### CODE REVIEW

**Comment by headius** [comment]
I love it! Will review and chat with other maintainers.

File: src/org/joni/Config.java [hunk 1]
```diff
@@ -22,54 +22,55 @@
...
-    final boolean USE_OP_PUSH_OR_JUMP_EXACT         = true;
-    final boolean USE_QTFR_PEEK_NEXT                = true;
+    final boolean USE_OP_PUSH_OR_JUMP_EXACT         = ConfigSupport.getBoolean("joni.use_op_push_or_jump_exact", true);
+    final boolean USE_QTFR_PEEK_NEXT                = ConfigSupport.getBoolean("joni.use_qtfr_peek_next", true);
 
-    final int INIT_MATCH_STACK_SIZE                 = 64;
+    final int INIT_MATCH_STACK_SIZE                 = ConfigSupport.getInt("joni.init_match_stack_size", 64);
 
-    final boolean DONT_OPTIMIZE                     = false;
+    final boolean DONT_OPTIMIZE                     = ConfigSupport.getBoolean("joni.dont_optimize", false);
```
Comment (headius) [inline src/org/joni/Config.java:63]: Let's just make this OPTIMIZE/joni.optimize=true. The double negative here makes my head hurt.

File: src/org/joni/ConfigSupport.java [hunk 2]
```diff
@@ -5,4 +5,9 @@ static boolean getBoolean(String property, boolean def) {
         String value = System.getProperty(property, def ? "true" : "false");
         return !value.equals("false");
     }
+
+  static int getInt(String property, int def) {
+    String value = System.getProperty(property);
+    return value != null ? Integer.parseInt(value) : def;
+  }
```
Comment (headius) [inline src/org/joni/ConfigSupport.java:12]: Please match 4-space indentation here.

File: src/org/joni/Search.java [hunk 3]
```diff
@@ -421,10 +421,10 @@ final int search(Matcher matcher, byte[]text, int textP, int textEnd, int textRa
             int targetP = regex.exactP;
             int targetEnd = regex.exactEnd;
 
-            int end, s, tlen1;
+            int end, s;
             int tail = targetEnd - 1;
+            int tlen1 = tail - targetP;
             if (USE_SUNDAY_QUICK_SEARCH) {
-                tlen1 = tail - targetP;
```
Comment (headius) [inline src/org/joni/Search.java:427]: Is this fixing a separate bug? If so we should move it to another PR because it's not config-related.

**Review by headius — CHANGES_REQUESTED** [review]
Looks great overall! Just a few minor changes needed.

File: src/org/joni/Search.java (hunk 3 above)
Comment (nezda) [reply src/org/joni/Search.java:427]: when I changed `USE_SUNDAY_QUICK_SEARCH ` to be initialized from system property it no longer compiled 

```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.0:compile (default-compile) on project joni: Compilation failure: Compilation failure:
[ERROR] /Users/nezda/code/joni.git/src/org/joni/Search.java:[438,59] variable tlen1 might not have been initialized
[ERROR] /Users/nezda/code/joni.git/src/org/joni/Search.java:[446,59] variable tlen1 might not have been initialized
```

so I tried to fix.

File: src/org/joni/Config.java (hunk 1 above)
Comment (nezda) [reply src/org/joni/Config.java]: I'm assuming the latter - pushed that, happy to adjust otherwise

Comment (headius) [reply src/org/joni/Config.java]: Yeah good call on preserving the public API. We might as well deprecate it too. I doubt anyone is using it.

File: src/org/joni/Search.java (hunk 3 above)
Comment (headius) [reply src/org/joni/Search.java:427]: I pinged @enebo to look at this too.

File: src/org/joni/Config.java (hunk 1 above)
Comment (nezda) [reply src/org/joni/Config.java]: 👍 deprecated it

**Comment by lopex** [comment]
I will disable constant propagations for those flags, I dont have an opinion on it though

File: src/org/joni/Search.java (hunk 3 above)
Comment (lopex) [reply src/org/joni/Search.java:427]: yeah, it was constant propagated since those interface fields are essentially static finals

**Comment by nezda** [comment]
> It will disable constant propagations for those flags, I don't have an opinion on it though

Seems like conservative thing to do is preserve whatever performance constant propagation gets us and just don't let `USE_SUNDAY_QUICK_SEARCH` be configurable (and revert my change to `Search.java). Doesn't preclude it being added in the future.

**Comment by lopex** [comment]
Oh, I was referring to all flags, those affecting parser shouldnt matter though.
But for example this one - USE_CEC (combination explostion check) was never used and tested and is a part of very hot code - opFail().
@headius, @enebo  some of the other flags are part of hot paths in stack manipulations and bytecode interpreter - do you think it will affect performance by much?

**Comment by nezda** [comment]
It is interesting an implicitly final variable initialized in that way stops being treated as a constant (but it seems clear something is going on related to that since code failed to compile without modification). I suppose final local variables could be introduce to ferry in values from those "constants" in that interface?

**Comment by lopex** [comment]
static finals where a direct translation for c #ifdef macros - I dont recall anything that could be a substitute for that in java.

**Comment by nezda** [comment]
I always thought final variables were basically equivalent to c #ifdef. I think this is my first counter-example. 

I could introduce a Configuration object with all final fields initialized in a constructor which would have benefit different instances within same jvm could have different configurations.

**Comment by headius** [comment]
Yeah, @lopex is right that this will prevent javac from eliminating code paths that were impossible to reach with literal static final values. That's why that sunday search code you patched suddenly started failing; with the original `true` value the uncompilable path was never followed and javac did not error.

I think the utility of making these configurable is worth any possible bytecode size increase here, even though it may tweak the inlining decisions of various JVMs. It should not generally affect performance when the same things inline as before, but some methods may cross size thresholds that prevent them from being trivially inlinable after this change.

If we could come up with a good way to audit which methods are affected by this, we could always move the conditional bodies out to separate static methods so they are small and don't impact the size of the original. There would be an additional inline that has to happen, so we'd have to weigh that against shrinking methods. We can make some determination by benchmarking expression matches that hit such methods heavily.

Personally, I would be surprised if this ends up being observable, and since there's no way to make these configurable without breaking the javac dead code optimization it may just be a price we have to accept (if it really has a cost).

**Comment by headius** [comment]
I would probably say we should just merge and then look at doing some benchmarking to find any performance regressions. We want these things to be configurable to make joni more flexible (to users other than JRuby also), so there may just be additional work necessary to find any slowdowns and deal with them.

**Comment by nezda** [comment]
For what it's worth, I concur with vote to merge this. I'd also be surprised if this had any significant negative performance consequence.

**Comment by headius** [comment]
Merged! We are debating some other fixes for release but otherwise we'll push this out soon.

**Comment by headius** [comment]
Moved to 2.2.0 because we're also landing some API changes (and this deprecates and changes some API as well).
